Update README for release 1.17 (#721)
kubernetes/client-go: README.md

### Versioning

- For each `v1.x.y` Kubernetes release, the major version (first digit)
would remain `0`.

- Bugfixes will result in the patch version (third digit) changing. PRs that are
cherry-picked into an older Kubernetes release branch will result in an update
to the corresponding branch in `client-go`, with a corresponding new tag

> changing the patch version.

Reviewer: just noticed the text below:
> A consequence of this is that `client-go` version numbers will be unrelated to
> Kubernetes version numbers.

That's not quite true now... they are related, as described above. I would drop that sentence as well.

Author: Done

### Versioning

- For each `v1.x.y` Kubernetes release, the major version (first digit)
would remain `0`.

> - Changes that add features in a backwards-compatible way will

Reviewer: This paragraph doesn't make a lot of sense any more, I would drop it.

Author: Dropped

### Versioning

> `client-go` follows [semver](http://semver.org/)-like tags.

Reviewer: drop "-like"

> `client-go` follows [semver](http://semver.org/)-like tags.

Reviewer: since we don't currently have the ability to increment major versions (xref https://github.com/kubernetes/kubernetes/issues/84372), I would limit this to describing the mapping from Kubernetes release to client-go release

#### Branches and tags.

We will create a new branch and tag for each increment in the minor version
number. We will create only a new tag for each increment in the patch
version number. See [semver](http://semver.org/) for definitions of major,
minor, and patch.

The HEAD of the master branch in client-go will track the HEAD of the master
branch in the main Kubernetes repo.

#### Compatibility: your code <-> client-go

> The `v0.x.y` tags indicate that incompatible changes can be expected at

Reviewer: ```suggestion
The `v0.x.y` tags indicate that go APIs may change in incompatible ways in
```

#### Branches and tags.

We will create a new branch and tag for each increment in the minor version
number. We will create only a new tag for each increment in the patch
version number. See [semver](http://semver.org/) for definitions of major,
minor, and patch.

The master branch will track HEAD in the main Kubernetes repo and
accumulate changes. Consider HEAD to have the version `x.(y+1).0-alpha`,

> where `x` and `y` are the current major and minor versions.

Reviewer: we are not adding major versions until we move to semantic import versioning. I would just say HEAD of master tracks Kubernetes master.

#### Compatibility: your code <-> client-go

> `client-go` follows [semver](http://semver.org/)-like tags, so until the minor

Reviewer: we don't have the ability to increment major versions without switching to semantic import versioning. The v0.x.y tags indicate that incompatible go changes can be expected at different versions.

version of client-go gets increased, your code will compile and will continue
to work with explicitly supported versions of Kubernetes clusters.

> You must use a dependency management system and pin a specific major version of

Reviewer: Drop most of this paragraph, major versions do not apply to client-go since all our releases are under the same v0 major version. Just leave the reference to INSTALL.md for guidelines on requiring a specific version of client-go

| Branch         | Canonical source code location       | Maintenance status            |
|----------------|--------------------------------------|-------------------------------|
| `release-1.4`  | Kubernetes main repo, 1.4 branch     | = -                           |
| `release-1.5`  | Kubernetes main repo, 1.5 branch     | = -                           |
| `release-2.0`  | Kubernetes main repo, 1.5 branch     | = -                           |
| `release-3.0`  | Kubernetes main repo, 1.6 branch     | = -                           |
| `release-4.0`  | Kubernetes main repo, 1.7 branch     | = -                           |
| `release-5.0`  | Kubernetes main repo, 1.8 branch     | = -                           |
| `release-6.0`  | Kubernetes main repo, 1.9 branch     | = -                           |
| `release-7.0`  | Kubernetes main repo, 1.10 branch    | = -                           |
| `release-8.0`  | Kubernetes main repo, 1.11 branch    | =-                            |
| `release-9.0`  | Kubernetes main repo, 1.12 branch    | =-                            |
| `release-10.0` | Kubernetes main repo, 1.13 branch    | =-                            |

> | `release-11.0` | Kubernetes main repo, 1.14 branch    | ✓                             |

Reviewer: this is about to not be maintained any more... do you want to drop this now or wait until the last 1.14.x release publishes later today?

Author: > do you want to drop this now or wait until the last 1.14.x release publishes later today?

I'd prefer to wait. I think it would make sense to turn down the godeps publishing-bot when CI for 1.14 is turned down as well, and if it takes O(month) from now (https://github.com/kubernetes/k8s.io/issues/500#issuecomment-564552627), it might be better to leave this here...

Author: As per the definition we are using here: ✓  = Changes in main Kubernetes repo are actively published to client-go by a bot
